File: flutter_digital_id_platform_interface/lib/src/models/digital_id_request_options.dart
import 'dart:convert';

import 'claim_path.dart';

/// Controls what data is requested from a digital identity document.
///
/// This enables selective disclosure: the caller can specify exactly which
/// claims are required (must be returned or the user cannot approve) versus
/// optional (the wallet/user may omit them).
///
/// The same options object works across iOS (PassKit descriptors), Android/Web
/// (OpenID4VP / DCQL), and the C# companion.
class DigitalIdRequestOptions {
  const DigitalIdRequestOptions({
    this.requiredClaims = const [],
    this.optionalClaims = const [],
    this.intentToRetain = false,
    this.nonce,
    this.allowMultipleCredentials = false,
    this.platformOptions,
  });

  /// Claims that must be present in the response. The user will not be able
  /// to approve the request unless all of these can be provided.
  ///
  /// Example (mdoc path style):
  /// ```dart
  /// ClaimPath(['org.iso.18013.5.1', 'family_name'])
  /// ```
  final List<ClaimPath> requiredClaims;

  /// Claims the caller would like but that are not mandatory for approval.
  final List<ClaimPath> optionalClaims;

  /// When true, the wallet should retain the disclosed values (subject to
  /// local privacy rules). Most privacy-respecting flows set this to false.
  final bool intentToRetain;

  /// Optional nonce for replay protection / binding to a specific session.
  final String? nonce;

  /// If true, the user may be offered multiple matching credentials (e.g.
  /// passport vs. driver's license) and can choose which one to present.
  ///
  /// NOTE: Currently passed through to native layers but full support depends
  /// on the platform implementation (Android DCQL can express this; iOS PassKit
  /// behavior depends on the entitlement + descriptors).
  final bool allowMultipleCredentials;

  /// Platform-specific escape hatch for advanced use cases (raw DCQL JSON,
  /// custom Apple descriptor tweaks, etc.). Use with caution — prefer the
  /// cross-platform fields above when possible.
  final Map<String, dynamic>? platformOptions;

  @override
  String toString() =>
      'DigitalIdRequestOptions(required: $requiredClaims, optional: $optionalClaims, retain: $intentToRetain)';

  /// Hand-coded serialization (no code generation).
  Map<String, dynamic> toMap() => {
        'requiredClaims': requiredClaims.map((e) => e.toMap()).toList(),
        'optionalClaims': optionalClaims.map((e) => e.toMap()).toList(),
        'intentToRetain': intentToRetain,
        'nonce': nonce,
        'allowMultipleCredentials': allowMultipleCredentials,
        'platformOptions': platformOptions,
      };

  factory DigitalIdRequestOptions.fromMap(Map<String, dynamic> map) {
    return DigitalIdRequestOptions(
      requiredClaims: (map['requiredClaims'] as List? ?? [])
          .map((e) => ClaimPath.fromMap(Map<String, dynamic>.from(e as Map)))
          .toList(),
      optionalClaims: (map['optionalClaims'] as List? ?? [])
          .map((e) => ClaimPath.fromMap(Map<String, dynamic>.from(e as Map)))
          .toList(),
      intentToRetain: map['intentToRetain'] as bool? ?? false,
      nonce: map['nonce'] as String?,
      allowMultipleCredentials: map['allowMultipleCredentials'] as bool? ?? false,
      platformOptions: map['platformOptions'] != null
          ? Map<String, dynamic>.from(map['platformOptions'] as Map)
          : null,
    );
  }

  String toJson() => jsonEncode(toMap());

  factory DigitalIdRequestOptions.fromJson(String source) =>
      DigitalIdRequestOptions.fromMap(jsonDecode(source) as Map<String, dynamic>);
}

File: DigitalId.Net/src/Models/DigitalIdRequestOptions.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace DigitalId;

/// <summary>
/// Controls what data is requested from a digital identity document (selective disclosure).
/// Matches Dart DigitalIdRequestOptions exactly for cross-platform use.
/// </summary>
public sealed record DigitalIdRequestOptions
{
    [JsonPropertyName("requiredClaims")]
    public IReadOnlyList<ClaimPath> RequiredClaims { get; init; } = [];

    [JsonPropertyName("optionalClaims")]
    public IReadOnlyList<ClaimPath> OptionalClaims { get; init; } = [];

    [JsonPropertyName("intentToRetain")]
    public bool IntentToRetain { get; init; }

    [JsonPropertyName("nonce")]
    public string? Nonce { get; init; }

    [JsonPropertyName("allowMultipleCredentials")]
    public bool AllowMultipleCredentials { get; init; }

    /// <summary>
    /// Platform-specific escape hatch (advanced use only).
    /// </summary>
    [JsonPropertyName("platformOptions")]
    public Dictionary<string, object?>? PlatformOptions { get; init; }

    /// <summary>
    /// Hand-coded serialization producing camelCase JSON matching Flutter toJson().
    /// </summary>
    public string ToJson() => JsonSerializer.Serialize(this, DigitalIdJsonContext.Default.DigitalIdRequestOptions);

    public static DigitalIdRequestOptions FromJson(string json) =>
        JsonSerializer.Deserialize(json, DigitalIdJsonContext.Default.DigitalIdRequestOptions) ?? throw new ArgumentException("Invalid DigitalIdRequestOptions JSON");

    /// <summary>
    /// Map using exact camelCase keys from Dart toMap().
    /// </summary>
    public Dictionary<string, object?> ToMap() => new()
    {
        ["requiredClaims"] = RequiredClaims.Select(c => c.ToMap()).ToList(),
        ["optionalClaims"] = OptionalClaims.Select(c => c.ToMap()).ToList(),
        ["intentToRetain"] = IntentToRetain,
        ["nonce"] = Nonce,
        ["allowMultipleCredentials"] = AllowMultipleCredentials,
        ["platformOptions"] = PlatformOptions
    };

    public static DigitalIdRequestOptions FromMap(Dictionary<string, object?> map)
    {
        var reqClaims = new List<ClaimPath>();
        if (map.TryGetValue("requiredClaims", out var req) && req is IEnumerable<object> reqList)
        {
            reqClaims = reqList
                .OfType<Dictionary<string, object?>>()
                .Select(ClaimPath.FromMap)
                .ToList();
        }

        var optClaims = new List<ClaimPath>();
        if (map.TryGetValue("optionalClaims", out var opt) && opt is IEnumerable<object> optList)
        {
            optClaims = optList
                .OfType<Dictionary<string, object?>>()
                .Select(ClaimPath.FromMap)
                .ToList();
        }

        return new DigitalIdRequestOptions
        {
            RequiredClaims = reqClaims,
            OptionalClaims = optClaims,
            IntentToRetain = map.TryGetValue("intentToRetain", out var ir) && ir is bool b && b,
            Nonce = map.TryGetValue("nonce", out var n) ? n?.ToString() : null,
            AllowMultipleCredentials = map.TryGetValue("allowMultipleCredentials", out var amc) && amc is bool b2 && b2,
            PlatformOptions = map.TryGetValue("platformOptions", out var po) && po is Dictionary<string, object?> pod ? pod : null
        };
    }
}
